controllers/templatesync/template_sync.go

```diff
@@ -866,3 +1058,73 @@ func getLatestStatusMessage(pol *policiesv1.Policy, tIndex int) string {
 
 	return tmplDetails.History[0].Message
 }
+
+func hasClusterwideFinalizer(pol *policiesv1.Policy) bool {
+	for _, finalizer := range pol.Finalizers {
+		if finalizer == utils.ClusterwideFinalizer {
+			return true
+		}
+	}
+
+	return false
+}
+
+// removeFinalizer iterates over the finalizers and removes the one specified from the policy.
+func removeFinalizer(pol *policiesv1.Policy, finalizer string) {
+	i := 0
+	finalizersLength := len(pol.Finalizers)
+
+	for i < finalizersLength {
+		if pol.Finalizers[i] == finalizer {
+			pol.Finalizers = append(pol.Finalizers[:i], pol.Finalizers[i+1:]...)
+			finalizersLength--
+		} else {
+			i++
+		}
+	}
+}
+
+// finalizerCleanup handles any steps required for cleaning up objects on the cluster prior to
+// removing the policy finalizer.
+func finalizerCleanup(
+	ctx context.Context, pol *policiesv1.Policy, rMapper meta.RESTMapper, dClient dynamic.Interface,
+) error {
+	for _, policyT := range pol.Spec.PolicyTemplates {
+		object, gvk, err := unstructured.UnstructuredJSONScheme.Decode(policyT.ObjectDefinition.Raw, nil, nil)
+		if err != nil {
+			return fmt.Errorf("failed to decode policy template with error: %w", err)
```

Comment:
Do you think we should `continue` here to handle the case where the user modifies the policy with a syntax error and then deletes it?

If not, then I think there should be a status set on the policy indicating that there is an invalid policy template so the user can fix it.

Reply:
Good point--for cleanup, we should probably try everything rather than returning. I'll implement an aggregated error here to gather all the errors together to return.